// markdown-component-loader/lib/string-replacement-cache.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _lowercaseHash = require('./lowercase-hash');

var _lowercaseHash2 = _interopRequireDefault(_lowercaseHash);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var noOpReplacer = function noOpReplacer(thing) {
  return thing;
};

var StringReplacementCache = function () {
  function StringReplacementCache(expression) {
    var outputReplacer = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : noOpReplacer;
    var identityReplacer = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : noOpReplacer;

    _classCallCheck(this, StringReplacementCache);

    this.expression = expression;
    this.outputReplacer = outputReplacer;
    this.identityReplacer = identityReplacer;
    this._cache = {};
  }

  _createClass(StringReplacementCache, [{
    key: 'load',
    value: function load(body) {
      var _this = this;

      var processed = body.replace(this.expression, function (match) {
        for (var _len = arguments.length, values = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
          values[_key - 1] = arguments[_key];
        }

        var identityHash = (0, _lowercaseHash2.default)(match);

        var identity = _this.identityReplacer.apply(_this, [identityHash, match].concat(values));

        _this._cache[identity] = _this.outputReplacer.apply(_this, [match].concat(values));

        return identity;
      });

      return processed;
    }
  }, {
    key: 'unload',
    value: function unload(body) {
      var _this2 = this;

      var processed = body;

      Object.keys(this._cache).forEach(function (identity) {
        return processed = processed.replace(new RegExp(identity.replace(/[-[\]/{}()*+?.\\^$|]/g, '\\$&'), 'g'), _this2._cache[identity]);
      });

      this._cache = {};
      return processed;
    }
  }]);

  return StringReplacementCache;
}();

exports.default = StringReplacementCache;
